Improve GMARE metric computation formula

Months ago I tried to improve GMARE to be compatible with DDP, but I had a very strange issue, and the distributed tests were failing, I thought I did something wrong (even I updated +10 metrics for DDP then) so I ignored it at that time, today I was trying to update it and send the PR but I had the same issue, the tests distributed tests are failing.
```
distributed_context_single_node_gloo = {'local_rank': 1, 'rank': 1, 'world_size': 2}

    @pytest.mark.distributed
    @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
    def test_distrib_gloo_cpu_or_gpu(distributed_context_single_node_gloo):
    
        device = idist.device()
>       _test_distrib_compute(device)


metric_device = device(type='cpu')

    def _test(metric_device):
        metric_device = torch.device(metric_device)
        m = GeometricMeanRelativeAbsoluteError(device=metric_device)
        torch.manual_seed(10 + rank)
    
        y_pred = torch.rand(size=(100,), device=device)
        y = torch.rand(size=(100,), device=device)
    
        m.update((y_pred, y))
    
        y_pred = idist.all_gather(y_pred)
        y = idist.all_gather(y)
    
        np_y = y.cpu().numpy()
        np_y_pred = y_pred.cpu().numpy()
    
        np_gmrae = np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
    
        # sum_y = 0
        # num_examples = 0
        # sum_of_errors = 0
        # np_gmrae = 0
        # sum_y += np_y.sum()
        # num_examples += np_y.shape[0]
        # y_mean = sum_y / num_examples
        # numerator = np.abs(y.view_as(y_pred) - y_pred)
        # denominator = np.abs(y.view_as(y_pred) - y_mean)
        # sum_of_errors += np.log(numerator / denominator).sum()
        # np_gmrae += np.exp((sum_of_errors / num_examples).mean())
    
>       assert m.compute() == pytest.approx(np_gmrae)
E       assert 1.3195266723632812 == 1.3320785760879517 ± 1.3e-06
E         +1.3195266723632812
E         -1.3320785760879517 ± 1.3e-06


FAILED tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py::test_distrib_gloo_cpu_or_gpu - assert 1.3195266723632812 == 1.3320785760879517 ± 1.3e-06
FAILED tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py::test_distrib_gloo_cpu_or_gpu - assert 1.3195266723632812 == 1.3320785760879517 ± 1.3e-06
```
And here is the code and the tests 
https://github.com/pytorch/ignite/compare/master...KickItLikeShika:improve-GMRAE-ddp

--- a/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
+++ b/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
@@ -1,10 +1,11 @@
-from typing import Tuple
+from typing import Tuple, cast
 
 import torch
 
+import ignite.distributed as idist
 from ignite.contrib.metrics.regression._base import _BaseRegression
 from ignite.exceptions import NotComputableError
-from ignite.metrics.metric import reinit__is_reduced, sync_all_reduce
+from ignite.metrics.metric import reinit__is_reduced
 
 
 class GeometricMeanRelativeAbsoluteError(_BaseRegression):
@@ -37,22 +38,45 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
 
     @reinit__is_reduced
     def reset(self) -> None:
-        # self._sum_y = 0.0  # type: Union[float, torch.Tensor]
-        self._num_examples = 0
-        self._sum_of_errors = torch.tensor(0.0, device=self._device)
+        self._predictions = []
+        self._targets = []
 
     def _update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
         y_pred, y = output[0].detach(), output[1].detach()
-        errors = torch.log(torch.abs(y - y_pred) / torch.abs(y - y.mean()))
-        self._sum_of_errors += torch.sum(errors).to(self._device)
-        self._num_examples += y.shape[0]
 
-        # np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
+        y_pred = y_pred.clone().to(self._device)
+        y = y.clone().to(self._device)
 
-    @sync_all_reduce("_sum_of_errors", "_num_examples")
+        self._predictions.append(y_pred)
+        self._targets.append(y)
+
     def compute(self) -> float:
-        if self._num_examples == 0:
+        if len(self._predictions) < 1 or len(self._targets) < 1:
             raise NotComputableError(
                 "GeometricMeanRelativeAbsoluteError must have at least one example before it can be computed."
             )
-        return torch.exp(self._sum_of_errors / self._num_examples).item()
+
+        _prediction_tensor = torch.cat(self._predictions, dim=0)
+        _target_tensor = torch.cat(self._targets, dim=0)
+
+        ws = idist.get_world_size()
+
+        if ws > 1:
+            # All gather across all processes
+            _prediction_tensor = cast(torch.Tensor, idist.all_gather(_prediction_tensor))
+            _target_tensor = cast(torch.Tensor, idist.all_gather(_target_tensor))
+
+        result = 0.0
+        if idist.get_rank() == 0:
+            # Run compute_fn on zero rank only
+            result = torch.exp(
+                torch.log(
+                    torch.abs(_target_tensor - _prediction_tensor) / torch.abs(_target_tensor - _target_tensor.mean())
+                ).mean()
+            ).item()
+
+        if ws > 1:
+            # broadcast result to all processes
+            result = cast(float, idist.broadcast(result, src=0))
+
+        return result

--- a/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
+++ b/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
@@ -1,4 +1,4 @@
-from typing import Tuple, cast
+from typing import List, Tuple, cast
 
 import torch
 
@@ -38,8 +38,8 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
 
     @reinit__is_reduced
     def reset(self) -> None:
-        self._predictions = []
-        self._targets = []
+        self._predictions = []  # type: List[torch.Tensor]
+        self._targets = []  # type: List[torch.Tensor]
 
     def _update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
         y_pred, y = output[0].detach(), output[1].detach()

--- a/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
+++ b/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
@@ -1,4 +1,4 @@
-from typing import Tuple, Union, cast
+from typing import Tuple
 
 import torch
 
@@ -37,18 +37,17 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
 
     @reinit__is_reduced
     def reset(self) -> None:
-        self._sum_y = 0.0  # type: Union[float, torch.Tensor]
+        # self._sum_y = 0.0  # type: Union[float, torch.Tensor]
         self._num_examples = 0
-        self._sum_of_errors = 0.0  # type: Union[float, torch.Tensor]
+        self._sum_of_errors = torch.tensor(0.0, device=self._device)
 
     def _update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
         y_pred, y = output[0].detach(), output[1].detach()
-        self._sum_y += y.sum()
+        errors = torch.log(torch.abs(y - y_pred) / torch.abs(y - y.mean()))
+        self._sum_of_errors += torch.sum(errors).to(self._device)
         self._num_examples += y.shape[0]
-        y_mean = self._sum_y / self._num_examples
-        numerator = torch.abs(y.view_as(y_pred) - y_pred)
-        denominator = torch.abs(y.view_as(y_pred) - y_mean)
-        self._sum_of_errors += torch.log(numerator / denominator).sum()
+
+        # np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
 
     @sync_all_reduce("_sum_of_errors", "_num_examples")
     def compute(self) -> float:
@@ -56,4 +55,4 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
             raise NotComputableError(
                 "GeometricMeanRelativeAbsoluteError must have at least one example before it can be computed."
             )
-        return torch.exp(torch.mean(cast(torch.Tensor, self._sum_of_errors) / self._num_examples)).item()
+        return torch.exp(self._sum_of_errors / self._num_examples).item()

--- a/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
+++ b/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
@@ -4,6 +4,7 @@ import torch
 
 from ignite.contrib.metrics.regression._base import _BaseRegression
 from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import reinit__is_reduced, sync_all_reduce
 
 
 class GeometricMeanRelativeAbsoluteError(_BaseRegression):
@@ -34,13 +35,14 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
             non-blocking. By default, CPU.
     """
 
+    @reinit__is_reduced
     def reset(self) -> None:
         self._sum_y = 0.0  # type: Union[float, torch.Tensor]
         self._num_examples = 0
         self._sum_of_errors = 0.0  # type: Union[float, torch.Tensor]
 
     def _update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
-        y_pred, y = output
+        y_pred, y = output[0].detach(), output[1].detach()
         self._sum_y += y.sum()
         self._num_examples += y.shape[0]
         y_mean = self._sum_y / self._num_examples
@@ -48,6 +50,7 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
         denominator = torch.abs(y.view_as(y_pred) - y_mean)
         self._sum_of_errors += torch.log(numerator / denominator).sum()
 
+    @sync_all_reduce("_sum_of_errors", "_num_examples")
     def compute(self) -> float:
         if self._num_examples == 0:
             raise NotComputableError(

--- a/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
+++ b/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
@@ -72,23 +72,14 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
         _prediction_tensor = torch.cat(self._predictions, dim=0)
         _target_tensor = torch.cat(self._targets, dim=0)
 
-        ws = idist.get_world_size()
+        # All gather across all processes
+        _prediction_tensor = cast(torch.Tensor, idist.all_gather(_prediction_tensor))
+        _target_tensor = cast(torch.Tensor, idist.all_gather(_target_tensor))
 
-        if ws > 1:
-            # All gather across all processes
-            _prediction_tensor = cast(torch.Tensor, idist.all_gather(_prediction_tensor))
-            _target_tensor = cast(torch.Tensor, idist.all_gather(_target_tensor))
-
-        result = 0.0
-        if idist.get_rank() == 0:
-            result = torch.exp(
-                torch.log(
-                    torch.abs(_target_tensor - _prediction_tensor) / torch.abs(_target_tensor - _target_tensor.mean())
-                ).mean()
-            ).item()
-
-        if ws > 1:
-            # broadcast result to all processes
-            result = cast(float, idist.broadcast(result, src=0))
+        result = torch.exp(
+            torch.log(
+                torch.abs(_target_tensor - _prediction_tensor) / torch.abs(_target_tensor - _target_tensor.mean())
+            ).mean()
+        ).item()
 
         return result

--- a/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
+++ b/ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py
@@ -14,7 +14,8 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
     .. math::
         \text{GMRAE} = \exp(\frac{1}{n}\sum_{j=1}^n \ln\frac{|A_j - P_j|}{|A_j - \bar{A}|})
 
-    where :math:`A_j` is the ground truth and :math:`P_j` is the predicted value.
+    where :math:`A_j` is the ground truth, :math:`P_j` is the predicted value
+    and :math: `bar{A}` is the mean of the ground truth.
 
     More details can be found in `Botchkarev 2018`__.
 
@@ -24,6 +25,18 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
     __ https://arxiv.org/abs/1809.03006
 
     Parameters are inherited from ``Metric.__init__``.
+
+    .. warning::
+
+        Current implementation of GMRAE stores all input data (output and target)
+        as tensors before computing the metric.
+        This can potentially lead to a memory error if the input data is larger than available RAM.
+
+        In distributed configuration, all stored data (output and target) is mutually collected across all processes
+        using all gather collective operation. This can potentially lead to a memory error.
+
+        Compute method compute the metric on zero rank process only and final result is broadcasted to
+        all processes.
 
     Args:
         output_transform: a callable that is used to transform the
@@ -68,7 +81,6 @@ class GeometricMeanRelativeAbsoluteError(_BaseRegression):
 
         result = 0.0
         if idist.get_rank() == 0:
-            # Run compute_fn on zero rank only
             result = torch.exp(
                 torch.log(
                     torch.abs(_target_tensor - _prediction_tensor) / torch.abs(_target_tensor - _target_tensor.mean())

--- a/tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py
+++ b/tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py
@@ -1,10 +1,10 @@
-import os
+# import os
 
 import numpy as np
 import pytest
 import torch
 
-import ignite.distributed as idist
+# import ignite.distributed as idist
 from ignite.contrib.metrics.regression import GeometricMeanRelativeAbsoluteError
 from ignite.engine import Engine
 from ignite.exceptions import NotComputableError
@@ -45,66 +45,12 @@ def test_geometric_mean_relative_absolute_error():
     assert np_gmrae == pytest.approx(m.compute())
 
 
-def test_geometric_mean_relative_absolute_error_2():
-
-    np.random.seed(1)
-    size = 105
-    np_y_pred = np.random.rand(size, 1)
-    np_y = np.random.rand(size, 1)
-    np.random.shuffle(np_y)
-
-    np_y_sum = 0
-    num_examples = 0
-    num_sum_of_errors = 0
-    np_gmrae = 0
-
-    m = GeometricMeanRelativeAbsoluteError()
-    y_pred = torch.from_numpy(np_y_pred)
-    y = torch.from_numpy(np_y)
-
-    m.reset()
-    n_iters = 15
-    batch_size = size // n_iters
-    for i in range(n_iters + 1):
-        idx = i * batch_size
-        np_y_i = np_y[idx : idx + batch_size]
-        np_y_pred_i = np_y_pred[idx : idx + batch_size]
-
-        np_y_sum += np_y_i.sum()
-        num_examples += np_y_i.shape[0]
-        np_mean = np_y_sum / num_examples
-
-        np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
-        m.update((y_pred[idx : idx + batch_size], y[idx : idx + batch_size]))
-
-    assert np.exp(np_gmrae / num_examples) == pytest.approx(m.compute())
-
-
 def test_integration_geometric_mean_relative_absolute_error():
 
-    np.random.seed(1)
-    size = 105
-    np_y_pred = np.random.rand(size, 1)
-    np_y = np.random.rand(size, 1)
-    np.random.shuffle(np_y)
+    y_pred = torch.rand(size=(100,))
+    y = torch.rand(size=(100,))
 
-    np_y_sum = 0
-    num_examples = 0
-    num_sum_of_errors = 0
-    np_gmrae = 0
-
-    n_iters = 15
-    batch_size = size // n_iters
-    for i in range(n_iters + 1):
-        idx = i * batch_size
-        np_y_i = np_y[idx : idx + batch_size]
-        np_y_pred_i = np_y_pred[idx : idx + batch_size]
-
-        np_y_sum += np_y_i.sum()
-        num_examples += np_y_i.shape[0]
-        np_mean = np_y_sum / num_examples
-
-        np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
+    batch_size = 10
 
     def update_fn(engine, batch):
         idx = (engine.state.iteration - 1) * batch_size
@@ -115,184 +61,269 @@ def test_integration_geometric_mean_relative_absolute_error():
     engine = Engine(update_fn)
 
     m = GeometricMeanRelativeAbsoluteError()
-    m.attach(engine, "geometric_mean_relative_absolute_error")
+    m.attach(engine, "gmrae")
 
-    data = list(range(size // batch_size))
-    gmrae = engine.run(data, max_epochs=1).metrics["geometric_mean_relative_absolute_error"]
+    np_y = y.numpy().ravel()
+    np_y_pred = y_pred.numpy().ravel()
 
-    assert np.exp(np_gmrae / num_examples) == pytest.approx(m.compute())
+    data = list(range(y_pred.shape[0] // batch_size))
+    gmrae = engine.run(data, max_epochs=1).metrics["gmrae"]
 
+    sum_errors = np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).sum()
+    np_len = len(y_pred)
+    np_ans = np.exp(sum_errors / np_len)
 
-def _test_distrib_compute(device):
-
-    rank = idist.get_rank()
-    torch.manual_seed(12)
-
-    def _test(metric_device):
-        metric_device = torch.device(metric_device)
-        m = GeometricMeanRelativeAbsoluteError(device=metric_device)
-        torch.manual_seed(10 + rank)
-
-        y_pred = torch.rand(size=(100,), device=device)
-        y = torch.rand(size=(100,), device=device)
-
-        m.update((y_pred, y))
-
-        y_pred = idist.all_gather(y_pred)
-        y = idist.all_gather(y)
-
-        np_y = y.cpu().numpy()
-        np_y_pred = y_pred.cpu().numpy()
-
-        np_gmrae = np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
-
-        # sum_y = 0
-        # num_examples = 0
-        # sum_of_errors = 0
-        # np_gmrae = 0
-        # sum_y += np_y.sum()
-        # num_examples += np_y.shape[0]
-        # y_mean = sum_y / num_examples
-        # numerator = np.abs(y.view_as(y_pred) - y_pred)
-        # denominator = np.abs(y.view_as(y_pred) - y_mean)
-        # sum_of_errors += np.log(numerator / denominator).sum()
-        # np_gmrae += np.exp((sum_of_errors / num_examples).mean())
-
-        assert m.compute() == pytest.approx(np_gmrae)
-
-    for _ in range(3):
-        _test("cpu")
-        if device.type != "xla":
-            _test(idist.device())
+    assert np_ans == pytest.approx(gmrae)
 
 
-def _test_distrib_integration(device):
+# def test_geometric_mean_relative_absolute_error_2():
 
-    rank = idist.get_rank()
-    torch.manual_seed(12)
+#     np.random.seed(1)
+#     size = 105
+#     np_y_pred = np.random.rand(size, 1)
+#     np_y = np.random.rand(size, 1)
+#     np.random.shuffle(np_y)
 
-    def _test(n_epochs, metric_device):
-        metric_device = torch.device(metric_device)
-        n_iters = 80
-        s = 16
-        n_classes = 2
+#     np_y_sum = 0
+#     num_examples = 0
+#     num_sum_of_errors = 0
+#     np_gmrae = 0
 
-        offset = n_iters * s
-        y_true = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
-        y_preds = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
+#     m = GeometricMeanRelativeAbsoluteError()
+#     y_pred = torch.from_numpy(np_y_pred)
+#     y = torch.from_numpy(np_y)
 
-        def update(engine, i):
-            return (
-                y_preds[i * s + rank * offset : (i + 1) * s + rank * offset],
-                y_true[i * s + rank * offset : (i + 1) * s + rank * offset],
-            )
+#     m.reset()
+#     n_iters = 15
+#     batch_size = size // n_iters
+#     for i in range(n_iters + 1):
+#         idx = i * batch_size
+#         np_y_i = np_y[idx : idx + batch_size]
+#         np_y_pred_i = np_y_pred[idx : idx + batch_size]
 
-        engine = Engine(update)
+#         np_y_sum += np_y_i.sum()
+#         num_examples += np_y_i.shape[0]
+#         np_mean = np_y_sum / num_examples
 
-        gmrae = GeometricMeanRelativeAbsoluteError(device=metric_device)
-        gmrae.attach(engine, "gmrae")
+#         np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
+#         m.update((y_pred[idx : idx + batch_size], y[idx : idx + batch_size]))
 
-        data = list(range(n_iters))
-        engine.run(data=data, max_epochs=n_epochs)
-
-        assert "gmrae" in engine.state.metrics
-
-        res = engine.state.metrics["gmrae"]
-
-        np_y = y_true.cpu().numpy()
-        np_y_pred = y_preds.cpu().numpy()
-
-        np_y_sum = 0
-        num_examples = 0
-        np_gmrae = 0
-        for i in range(n_iters + 1):
-            np_y_i = np_y[i * s + rank * offset : (i + 1) * s + rank * offset]
-            np_y_pred_i = np_y_pred[i * s + rank * offset : (i + 1) * s + rank * offset]
-
-            np_y_sum += np_y_i.sum()
-            num_examples += np_y_i.shape[0]
-            np_mean = np_y_sum / num_examples
-
-            np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
-
-        assert pytest.approx(res, rel=1e-4) == np.exp(np_gmrae / num_examples)
-
-    metric_devices = ["cpu"]
-    if device.type != "xla":
-        metric_devices.append(idist.device())
-    for metric_device in metric_devices:
-        for _ in range(2):
-            _test(n_epochs=1, metric_device=metric_device)
-            _test(n_epochs=2, metric_device=metric_device)
+#     assert np.exp(np_gmrae / num_examples) == pytest.approx(m.compute())
 
 
-@pytest.mark.distributed
-@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
-@pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU")
-def test_distrib_nccl_gpu(distributed_context_single_node_nccl):
+# def test_integration_geometric_mean_relative_absolute_error():
 
-    device = idist.device()
-    _test_distrib_compute(device)
-    _test_distrib_integration(device)
+#     np.random.seed(1)
+#     size = 105
+#     np_y_pred = np.random.rand(size, 1)
+#     np_y = np.random.rand(size, 1)
+#     np.random.shuffle(np_y)
 
+#     np_y_sum = 0
+#     num_examples = 0
+#     num_sum_of_errors = 0
+#     np_gmrae = 0
 
-@pytest.mark.distributed
-@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
-def test_distrib_gloo_cpu_or_gpu(distributed_context_single_node_gloo):
+#     n_iters = 15
+#     batch_size = size // n_iters
+#     for i in range(n_iters + 1):
+#         idx = i * batch_size
+#         np_y_i = np_y[idx : idx + batch_size]
+#         np_y_pred_i = np_y_pred[idx : idx + batch_size]
 
-    device = idist.device()
-    _test_distrib_compute(device)
-    _test_distrib_integration(device)
+#         np_y_sum += np_y_i.sum()
+#         num_examples += np_y_i.shape[0]
+#         np_mean = np_y_sum / num_examples
 
+#         np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
 
-@pytest.mark.distributed
-@pytest.mark.skipif(not idist.has_hvd_support, reason="Skip if no Horovod dist support")
-@pytest.mark.skipif("WORLD_SIZE" in os.environ, reason="Skip if launched as multiproc")
-def test_distrib_hvd(gloo_hvd_executor):
-    device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
-    nproc = 4 if not torch.cuda.is_available() else torch.cuda.device_count()
+#     def update_fn(engine, batch):
+#         idx = (engine.state.iteration - 1) * batch_size
+#         y_true_batch = np_y[idx : idx + batch_size]
+#         y_pred_batch = np_y_pred[idx : idx + batch_size]
+#         return torch.from_numpy(y_pred_batch), torch.from_numpy(y_true_batch)
 
-    gloo_hvd_executor(_test_distrib_compute, (device,), np=nproc, do_init=True)
-    gloo_hvd_executor(_test_distrib_integration, (device,), np=nproc, do_init=True)
+#     engine = Engine(update_fn)
 
+#     m = GeometricMeanRelativeAbsoluteError()
+#     m.attach(engine, "geometric_mean_relative_absolute_error")
 
-@pytest.mark.multinode_distributed
-@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
-@pytest.mark.skipif("MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
-def test_multinode_distrib_gloo_cpu_or_gpu(distributed_context_multi_node_gloo):
+#     data = list(range(size // batch_size))
+#     gmrae = engine.run(data, max_epochs=1).metrics["geometric_mean_relative_absolute_error"]
 
-    device = idist.device()
-    _test_distrib_compute(device)
-    _test_distrib_integration(device)
-
-
-@pytest.mark.multinode_distributed
-@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
-@pytest.mark.skipif("GPU_MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
-def test_multinode_distrib_nccl_gpu(distributed_context_multi_node_nccl):
-
-    device = idist.device()
-    _test_distrib_compute(device)
+#     assert np.exp(np_gmrae / num_examples) == pytest.approx(m.compute())
 
 
-@pytest.mark.tpu
-@pytest.mark.skipif("NUM_TPU_WORKERS" in os.environ, reason="Skip if NUM_TPU_WORKERS is in env vars")
-@pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
-def test_distrib_single_device_xla():
-    device = idist.device()
-    _test_distrib_compute(device)
-    _test_distrib_integration(device)
+# def _test_distrib_compute(device):
+
+#     rank = idist.get_rank()
+#     torch.manual_seed(12)
+
+#     def _test(metric_device):
+#         metric_device = torch.device(metric_device)
+#         m = GeometricMeanRelativeAbsoluteError(device=metric_device)
+#         torch.manual_seed(10 + rank)
+
+#         y_pred = torch.rand(size=(100,), device=device)
+#         y = torch.rand(size=(100,), device=device)
+
+#         m.update((y_pred, y))
+
+#         y_pred = idist.all_gather(y_pred)
+#         y = idist.all_gather(y)
+
+#         np_y = y.cpu().numpy()
+#         np_y_pred = y_pred.cpu().numpy()
+
+#         np_gmrae = np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
+
+#         # sum_y = 0
+#         # num_examples = 0
+#         # sum_of_errors = 0
+#         # np_gmrae = 0
+#         # sum_y += np_y.sum()
+#         # num_examples += np_y.shape[0]
+#         # y_mean = sum_y / num_examples
+#         # numerator = np.abs(y.view_as(y_pred) - y_pred)
+#         # denominator = np.abs(y.view_as(y_pred) - y_mean)
+#         # sum_of_errors += np.log(numerator / denominator).sum()
+#         # np_gmrae += np.exp((sum_of_errors / num_examples).mean())
+
+#         assert m.compute() == pytest.approx(np_gmrae)
+
+#     for _ in range(3):
+#         _test("cpu")
+#         if device.type != "xla":
+#             _test(idist.device())
 
 
-def _test_distrib_xla_nprocs(index):
-    device = idist.device()
-    _test_distrib_compute(device)
-    _test_distrib_integration(device)
+# def _test_distrib_integration(device):
+
+#     rank = idist.get_rank()
+#     torch.manual_seed(12)
+
+#     def _test(n_epochs, metric_device):
+#         metric_device = torch.device(metric_device)
+#         n_iters = 80
+#         s = 16
+#         n_classes = 2
+
+#         offset = n_iters * s
+#         y_true = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
+#         y_preds = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
+
+#         def update(engine, i):
+#             return (
+#                 y_preds[i * s + rank * offset : (i + 1) * s + rank * offset],
+#                 y_true[i * s + rank * offset : (i + 1) * s + rank * offset],
+#             )
+
+#         engine = Engine(update)
+
+#         gmrae = GeometricMeanRelativeAbsoluteError(device=metric_device)
+#         gmrae.attach(engine, "gmrae")
+
+#         data = list(range(n_iters))
+#         engine.run(data=data, max_epochs=n_epochs)
+
+#         assert "gmrae" in engine.state.metrics
+
+#         res = engine.state.metrics["gmrae"]
+
+#         np_y = y_true.cpu().numpy()
+#         np_y_pred = y_preds.cpu().numpy()
+
+#         np_y_sum = 0
+#         num_examples = 0
+#         np_gmrae = 0
+#         for i in range(n_iters + 1):
+#             np_y_i = np_y[i * s + rank * offset : (i + 1) * s + rank * offset]
+#             np_y_pred_i = np_y_pred[i * s + rank * offset : (i + 1) * s + rank * offset]
+
+#             np_y_sum += np_y_i.sum()
+#             num_examples += np_y_i.shape[0]
+#             np_mean = np_y_sum / num_examples
+
+#             np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
+
+#         assert pytest.approx(res, rel=1e-4) == np.exp(np_gmrae / num_examples)
+
+#     metric_devices = ["cpu"]
+#     if device.type != "xla":
+#         metric_devices.append(idist.device())
+#     for metric_device in metric_devices:
+#         for _ in range(2):
+#             _test(n_epochs=1, metric_device=metric_device)
+#             _test(n_epochs=2, metric_device=metric_device)
 
 
-@pytest.mark.tpu
-@pytest.mark.skipif("NUM_TPU_WORKERS" not in os.environ, reason="Skip if no NUM_TPU_WORKERS in env vars")
-@pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
-def test_distrib_xla_nprocs(xmp_executor):
-    n = int(os.environ["NUM_TPU_WORKERS"])
-    xmp_executor(_test_distrib_xla_nprocs, args=(), nprocs=n)
+# @pytest.mark.distributed
+# @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+# @pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU")
+# def test_distrib_nccl_gpu(distributed_context_single_node_nccl):
+
+#     device = idist.device()
+#     _test_distrib_compute(device)
+#     _test_distrib_integration(device)
+
+
+# @pytest.mark.distributed
+# @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+# def test_distrib_gloo_cpu_or_gpu(distributed_context_single_node_gloo):
+
+#     device = idist.device()
+#     _test_distrib_compute(device)
+#     _test_distrib_integration(device)
+
+
+# @pytest.mark.distributed
+# @pytest.mark.skipif(not idist.has_hvd_support, reason="Skip if no Horovod dist support")
+# @pytest.mark.skipif("WORLD_SIZE" in os.environ, reason="Skip if launched as multiproc")
+# def test_distrib_hvd(gloo_hvd_executor):
+#     device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
+#     nproc = 4 if not torch.cuda.is_available() else torch.cuda.device_count()
+
+#     gloo_hvd_executor(_test_distrib_compute, (device,), np=nproc, do_init=True)
+#     gloo_hvd_executor(_test_distrib_integration, (device,), np=nproc, do_init=True)
+
+
+# @pytest.mark.multinode_distributed
+# @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+# @pytest.mark.skipif("MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
+# def test_multinode_distrib_gloo_cpu_or_gpu(distributed_context_multi_node_gloo):
+
+#     device = idist.device()
+#     _test_distrib_compute(device)
+#     _test_distrib_integration(device)
+
+
+# @pytest.mark.multinode_distributed
+# @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+# @pytest.mark.skipif("GPU_MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
+# def test_multinode_distrib_nccl_gpu(distributed_context_multi_node_nccl):
+
+#     device = idist.device()
+#     _test_distrib_compute(device)
+
+
+# @pytest.mark.tpu
+# @pytest.mark.skipif("NUM_TPU_WORKERS" in os.environ, reason="Skip if NUM_TPU_WORKERS is in env vars")
+# @pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
+# def test_distrib_single_device_xla():
+#     device = idist.device()
+#     _test_distrib_compute(device)
+#     _test_distrib_integration(device)
+
+
+# def _test_distrib_xla_nprocs(index):
+#     device = idist.device()
+#     _test_distrib_compute(device)
+#     _test_distrib_integration(device)
+
+
+# @pytest.mark.tpu
+# @pytest.mark.skipif("NUM_TPU_WORKERS" not in os.environ, reason="Skip if no NUM_TPU_WORKERS in env vars")
+# @pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
+# def test_distrib_xla_nprocs(xmp_executor):
+#     n = int(os.environ["NUM_TPU_WORKERS"])
+#     xmp_executor(_test_distrib_xla_nprocs, args=(), nprocs=n)

--- a/tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py
+++ b/tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py
@@ -1,7 +1,10 @@
+import os
+
 import numpy as np
 import pytest
 import torch
 
+import ignite.distributed as idist
 from ignite.contrib.metrics.regression import GeometricMeanRelativeAbsoluteError
 from ignite.engine import Engine
 from ignite.exceptions import NotComputableError
@@ -77,7 +80,7 @@ def test_geometric_mean_relative_absolute_error_2():
     assert np.exp(np_gmrae / num_examples) == pytest.approx(m.compute())
 
 
-def test_integration_geometric_mean_relative_absolute_error_with_output_transform():
+def test_integration_geometric_mean_relative_absolute_error():
 
     np.random.seed(1)
     size = 105
@@ -107,14 +110,189 @@ def test_integration_geometric_mean_relative_absolute_error_with_output_transfor
         idx = (engine.state.iteration - 1) * batch_size
         y_true_batch = np_y[idx : idx + batch_size]
         y_pred_batch = np_y_pred[idx : idx + batch_size]
-        return idx, torch.from_numpy(y_pred_batch), torch.from_numpy(y_true_batch)
+        return torch.from_numpy(y_pred_batch), torch.from_numpy(y_true_batch)
 
     engine = Engine(update_fn)
 
-    m = GeometricMeanRelativeAbsoluteError(output_transform=lambda x: (x[1], x[2]))
+    m = GeometricMeanRelativeAbsoluteError()
     m.attach(engine, "geometric_mean_relative_absolute_error")
 
     data = list(range(size // batch_size))
     gmrae = engine.run(data, max_epochs=1).metrics["geometric_mean_relative_absolute_error"]
 
     assert np.exp(np_gmrae / num_examples) == pytest.approx(m.compute())
+
+
+def _test_distrib_compute(device):
+
+    rank = idist.get_rank()
+    torch.manual_seed(12)
+
+    def _test(metric_device):
+        metric_device = torch.device(metric_device)
+        m = GeometricMeanRelativeAbsoluteError(device=metric_device)
+        torch.manual_seed(10 + rank)
+
+        y_pred = torch.rand(size=(100,), device=device)
+        y = torch.rand(size=(100,), device=device)
+
+        m.update((y_pred, y))
+
+        y_pred = idist.all_gather(y_pred)
+        y = idist.all_gather(y)
+
+        np_y = y.cpu().numpy()
+        np_y_pred = y_pred.cpu().numpy()
+
+        np_gmrae = np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
+
+        # sum_y = 0
+        # num_examples = 0
+        # sum_of_errors = 0
+        # np_gmrae = 0
+        # sum_y += np_y.sum()
+        # num_examples += np_y.shape[0]
+        # y_mean = sum_y / num_examples
+        # numerator = np.abs(y.view_as(y_pred) - y_pred)
+        # denominator = np.abs(y.view_as(y_pred) - y_mean)
+        # sum_of_errors += np.log(numerator / denominator).sum()
+        # np_gmrae += np.exp((sum_of_errors / num_examples).mean())
+
+        assert m.compute() == pytest.approx(np_gmrae)
+
+    for _ in range(3):
+        _test("cpu")
+        if device.type != "xla":
+            _test(idist.device())
+
+
+def _test_distrib_integration(device):
+
+    rank = idist.get_rank()
+    torch.manual_seed(12)
+
+    def _test(n_epochs, metric_device):
+        metric_device = torch.device(metric_device)
+        n_iters = 80
+        s = 16
+        n_classes = 2
+
+        offset = n_iters * s
+        y_true = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
+        y_preds = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
+
+        def update(engine, i):
+            return (
+                y_preds[i * s + rank * offset : (i + 1) * s + rank * offset],
+                y_true[i * s + rank * offset : (i + 1) * s + rank * offset],
+            )
+
+        engine = Engine(update)
+
+        gmrae = GeometricMeanRelativeAbsoluteError(device=metric_device)
+        gmrae.attach(engine, "gmrae")
+
+        data = list(range(n_iters))
+        engine.run(data=data, max_epochs=n_epochs)
+
+        assert "gmrae" in engine.state.metrics
+
+        res = engine.state.metrics["gmrae"]
+
+        np_y = y_true.cpu().numpy()
+        np_y_pred = y_preds.cpu().numpy()
+
+        np_y_sum = 0
+        num_examples = 0
+        np_gmrae = 0
+        for i in range(n_iters + 1):
+            np_y_i = np_y[i * s + rank * offset : (i + 1) * s + rank * offset]
+            np_y_pred_i = np_y_pred[i * s + rank * offset : (i + 1) * s + rank * offset]
+
+            np_y_sum += np_y_i.sum()
+            num_examples += np_y_i.shape[0]
+            np_mean = np_y_sum / num_examples
+
+            np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
+
+        assert pytest.approx(res, rel=1e-4) == np.exp(np_gmrae / num_examples)
+
+    metric_devices = ["cpu"]
+    if device.type != "xla":
+        metric_devices.append(idist.device())
+    for metric_device in metric_devices:
+        for _ in range(2):
+            _test(n_epochs=1, metric_device=metric_device)
+            _test(n_epochs=2, metric_device=metric_device)
+
+
+@pytest.mark.distributed
+@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+@pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU")
+def test_distrib_nccl_gpu(distributed_context_single_node_nccl):
+
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
+
+
+@pytest.mark.distributed
+@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+def test_distrib_gloo_cpu_or_gpu(distributed_context_single_node_gloo):
+
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
+
+
+@pytest.mark.distributed
+@pytest.mark.skipif(not idist.has_hvd_support, reason="Skip if no Horovod dist support")
+@pytest.mark.skipif("WORLD_SIZE" in os.environ, reason="Skip if launched as multiproc")
+def test_distrib_hvd(gloo_hvd_executor):
+    device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
+    nproc = 4 if not torch.cuda.is_available() else torch.cuda.device_count()
+
+    gloo_hvd_executor(_test_distrib_compute, (device,), np=nproc, do_init=True)
+    gloo_hvd_executor(_test_distrib_integration, (device,), np=nproc, do_init=True)
+
+
+@pytest.mark.multinode_distributed
+@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+@pytest.mark.skipif("MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
+def test_multinode_distrib_gloo_cpu_or_gpu(distributed_context_multi_node_gloo):
+
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
+
+
+@pytest.mark.multinode_distributed
+@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+@pytest.mark.skipif("GPU_MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
+def test_multinode_distrib_nccl_gpu(distributed_context_multi_node_nccl):
+
+    device = idist.device()
+    _test_distrib_compute(device)
+
+
+@pytest.mark.tpu
+@pytest.mark.skipif("NUM_TPU_WORKERS" in os.environ, reason="Skip if NUM_TPU_WORKERS is in env vars")
+@pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
+def test_distrib_single_device_xla():
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
+
+
+def _test_distrib_xla_nprocs(index):
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
+
+
+@pytest.mark.tpu
+@pytest.mark.skipif("NUM_TPU_WORKERS" not in os.environ, reason="Skip if no NUM_TPU_WORKERS in env vars")
+@pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
+def test_distrib_xla_nprocs(xmp_executor):
+    n = int(os.environ["NUM_TPU_WORKERS"])
+    xmp_executor(_test_distrib_xla_nprocs, args=(), nprocs=n)

--- a/tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py
+++ b/tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py
@@ -99,7 +99,7 @@ def _test_distrib_compute(device):
 
         np_gmrae = np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
 
-        assert m.compute() == pytest.approx(np_gmrae)
+        assert m.compute() == pytest.approx(np_gmrae, rel=1e-4)
 
     for _ in range(3):
         _test("cpu")

--- a/tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py
+++ b/tests/ignite/contrib/metrics/regression/test_geometric_mean_relative_absolute_error.py
@@ -1,10 +1,10 @@
-# import os
+import os
 
 import numpy as np
 import pytest
 import torch
 
-# import ignite.distributed as idist
+import ignite.distributed as idist
 from ignite.contrib.metrics.regression import GeometricMeanRelativeAbsoluteError
 from ignite.engine import Engine
 from ignite.exceptions import NotComputableError
@@ -29,7 +29,7 @@ def test_wrong_input_shapes():
         m.update((torch.rand(4, 1), torch.rand(4,)))
 
 
-def test_geometric_mean_relative_absolute_error():
+def test_compute():
     size = 51
     np_y_pred = np.random.rand(size,)
     np_y = np.random.rand(size,)
@@ -45,7 +45,7 @@ def test_geometric_mean_relative_absolute_error():
     assert np_gmrae == pytest.approx(m.compute())
 
 
-def test_integration_geometric_mean_relative_absolute_error():
+def test_integration():
 
     y_pred = torch.rand(size=(100,))
     y = torch.rand(size=(100,))
@@ -76,254 +76,153 @@ def test_integration_geometric_mean_relative_absolute_error():
     assert np_ans == pytest.approx(gmrae)
 
 
-# def test_geometric_mean_relative_absolute_error_2():
+def _test_distrib_compute(device):
 
-#     np.random.seed(1)
-#     size = 105
-#     np_y_pred = np.random.rand(size, 1)
-#     np_y = np.random.rand(size, 1)
-#     np.random.shuffle(np_y)
+    rank = idist.get_rank()
+    torch.manual_seed(12)
 
-#     np_y_sum = 0
-#     num_examples = 0
-#     num_sum_of_errors = 0
-#     np_gmrae = 0
+    def _test(metric_device):
+        metric_device = torch.device(metric_device)
+        m = GeometricMeanRelativeAbsoluteError(device=metric_device)
+        torch.manual_seed(10 + rank)
 
-#     m = GeometricMeanRelativeAbsoluteError()
-#     y_pred = torch.from_numpy(np_y_pred)
-#     y = torch.from_numpy(np_y)
+        y_pred = torch.rand(size=(100,), device=device)
+        y = torch.rand(size=(100,), device=device)
 
-#     m.reset()
-#     n_iters = 15
-#     batch_size = size // n_iters
-#     for i in range(n_iters + 1):
-#         idx = i * batch_size
-#         np_y_i = np_y[idx : idx + batch_size]
-#         np_y_pred_i = np_y_pred[idx : idx + batch_size]
+        m.update((y_pred, y))
 
-#         np_y_sum += np_y_i.sum()
-#         num_examples += np_y_i.shape[0]
-#         np_mean = np_y_sum / num_examples
+        y_pred = idist.all_gather(y_pred)
+        y = idist.all_gather(y)
 
-#         np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
-#         m.update((y_pred[idx : idx + batch_size], y[idx : idx + batch_size]))
+        np_y = y.cpu().numpy()
+        np_y_pred = y_pred.cpu().numpy()
 
-#     assert np.exp(np_gmrae / num_examples) == pytest.approx(m.compute())
+        np_gmrae = np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
 
+        assert m.compute() == pytest.approx(np_gmrae)
 
-# def test_integration_geometric_mean_relative_absolute_error():
+    for _ in range(3):
+        _test("cpu")
+        if device.type != "xla":
+            _test(idist.device())
 
-#     np.random.seed(1)
-#     size = 105
-#     np_y_pred = np.random.rand(size, 1)
-#     np_y = np.random.rand(size, 1)
-#     np.random.shuffle(np_y)
 
-#     np_y_sum = 0
-#     num_examples = 0
-#     num_sum_of_errors = 0
-#     np_gmrae = 0
+def _test_distrib_integration(device):
 
-#     n_iters = 15
-#     batch_size = size // n_iters
-#     for i in range(n_iters + 1):
-#         idx = i * batch_size
-#         np_y_i = np_y[idx : idx + batch_size]
-#         np_y_pred_i = np_y_pred[idx : idx + batch_size]
+    rank = idist.get_rank()
+    torch.manual_seed(12)
 
-#         np_y_sum += np_y_i.sum()
-#         num_examples += np_y_i.shape[0]
-#         np_mean = np_y_sum / num_examples
+    def _test(n_epochs, metric_device):
+        metric_device = torch.device(metric_device)
+        n_iters = 80
+        s = 16
+        n_classes = 2
 
-#         np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
+        offset = n_iters * s
+        y_true = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
+        y_preds = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
 
-#     def update_fn(engine, batch):
-#         idx = (engine.state.iteration - 1) * batch_size
-#         y_true_batch = np_y[idx : idx + batch_size]
-#         y_pred_batch = np_y_pred[idx : idx + batch_size]
-#         return torch.from_numpy(y_pred_batch), torch.from_numpy(y_true_batch)
+        def update(engine, i):
+            return (
+                y_preds[i * s + rank * offset : (i + 1) * s + rank * offset],
+                y_true[i * s + rank * offset : (i + 1) * s + rank * offset],
+            )
 
-#     engine = Engine(update_fn)
+        engine = Engine(update)
 
-#     m = GeometricMeanRelativeAbsoluteError()
-#     m.attach(engine, "geometric_mean_relative_absolute_error")
+        gmrae = GeometricMeanRelativeAbsoluteError(device=metric_device)
+        gmrae.attach(engine, "gmrae")
 
-#     data = list(range(size // batch_size))
-#     gmrae = engine.run(data, max_epochs=1).metrics["geometric_mean_relative_absolute_error"]
+        data = list(range(n_iters))
+        engine.run(data=data, max_epochs=n_epochs)
 
-#     assert np.exp(np_gmrae / num_examples) == pytest.approx(m.compute())
+        assert "gmrae" in engine.state.metrics
 
+        res = engine.state.metrics["gmrae"]
 
-# def _test_distrib_compute(device):
+        np_y = y_true.cpu().numpy()
+        np_y_pred = y_preds.cpu().numpy()
 
-#     rank = idist.get_rank()
-#     torch.manual_seed(12)
+        np_gmrae = np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
 
-#     def _test(metric_device):
-#         metric_device = torch.device(metric_device)
-#         m = GeometricMeanRelativeAbsoluteError(device=metric_device)
-#         torch.manual_seed(10 + rank)
+        assert pytest.approx(res, rel=1e-4) == np_gmrae
 
-#         y_pred = torch.rand(size=(100,), device=device)
-#         y = torch.rand(size=(100,), device=device)
+    metric_devices = ["cpu"]
+    if device.type != "xla":
+        metric_devices.append(idist.device())
+    for metric_device in metric_devices:
+        for _ in range(2):
+            _test(n_epochs=1, metric_device=metric_device)
+            _test(n_epochs=2, metric_device=metric_device)
 
-#         m.update((y_pred, y))
 
-#         y_pred = idist.all_gather(y_pred)
-#         y = idist.all_gather(y)
+@pytest.mark.distributed
+@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+@pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU")
+def test_distrib_nccl_gpu(distributed_context_single_node_nccl):
 
-#         np_y = y.cpu().numpy()
-#         np_y_pred = y_pred.cpu().numpy()
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
 
-#         np_gmrae = np.exp(np.log(np.abs(np_y - np_y_pred) / np.abs(np_y - np_y.mean())).mean())
 
-#         # sum_y = 0
-#         # num_examples = 0
-#         # sum_of_errors = 0
-#         # np_gmrae = 0
-#         # sum_y += np_y.sum()
-#         # num_examples += np_y.shape[0]
-#         # y_mean = sum_y / num_examples
-#         # numerator = np.abs(y.view_as(y_pred) - y_pred)
-#         # denominator = np.abs(y.view_as(y_pred) - y_mean)
-#         # sum_of_errors += np.log(numerator / denominator).sum()
-#         # np_gmrae += np.exp((sum_of_errors / num_examples).mean())
+@pytest.mark.distributed
+@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+def test_distrib_gloo_cpu_or_gpu(distributed_context_single_node_gloo):
 
-#         assert m.compute() == pytest.approx(np_gmrae)
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
 
-#     for _ in range(3):
-#         _test("cpu")
-#         if device.type != "xla":
-#             _test(idist.device())
 
+@pytest.mark.distributed
+@pytest.mark.skipif(not idist.has_hvd_support, reason="Skip if no Horovod dist support")
+@pytest.mark.skipif("WORLD_SIZE" in os.environ, reason="Skip if launched as multiproc")
+def test_distrib_hvd(gloo_hvd_executor):
+    device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
+    nproc = 4 if not torch.cuda.is_available() else torch.cuda.device_count()
 
-# def _test_distrib_integration(device):
+    gloo_hvd_executor(_test_distrib_compute, (device,), np=nproc, do_init=True)
+    gloo_hvd_executor(_test_distrib_integration, (device,), np=nproc, do_init=True)
 
-#     rank = idist.get_rank()
-#     torch.manual_seed(12)
 
-#     def _test(n_epochs, metric_device):
-#         metric_device = torch.device(metric_device)
-#         n_iters = 80
-#         s = 16
-#         n_classes = 2
+@pytest.mark.multinode_distributed
+@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+@pytest.mark.skipif("MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
+def test_multinode_distrib_gloo_cpu_or_gpu(distributed_context_multi_node_gloo):
 
-#         offset = n_iters * s
-#         y_true = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
-#         y_preds = torch.rand(size=(offset * idist.get_world_size(),)).to(device)
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
 
-#         def update(engine, i):
-#             return (
-#                 y_preds[i * s + rank * offset : (i + 1) * s + rank * offset],
-#                 y_true[i * s + rank * offset : (i + 1) * s + rank * offset],
-#             )
 
-#         engine = Engine(update)
+@pytest.mark.multinode_distributed
+@pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
+@pytest.mark.skipif("GPU_MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
+def test_multinode_distrib_nccl_gpu(distributed_context_multi_node_nccl):
 
-#         gmrae = GeometricMeanRelativeAbsoluteError(device=metric_device)
-#         gmrae.attach(engine, "gmrae")
+    device = idist.device()
+    _test_distrib_compute(device)
 
-#         data = list(range(n_iters))
-#         engine.run(data=data, max_epochs=n_epochs)
 
-#         assert "gmrae" in engine.state.metrics
+@pytest.mark.tpu
+@pytest.mark.skipif("NUM_TPU_WORKERS" in os.environ, reason="Skip if NUM_TPU_WORKERS is in env vars")
+@pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
+def test_distrib_single_device_xla():
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
 
-#         res = engine.state.metrics["gmrae"]
 
-#         np_y = y_true.cpu().numpy()
-#         np_y_pred = y_preds.cpu().numpy()
+def _test_distrib_xla_nprocs(index):
+    device = idist.device()
+    _test_distrib_compute(device)
+    _test_distrib_integration(device)
 
-#         np_y_sum = 0
-#         num_examples = 0
-#         np_gmrae = 0
-#         for i in range(n_iters + 1):
-#             np_y_i = np_y[i * s + rank * offset : (i + 1) * s + rank * offset]
-#             np_y_pred_i = np_y_pred[i * s + rank * offset : (i + 1) * s + rank * offset]
 
-#             np_y_sum += np_y_i.sum()
-#             num_examples += np_y_i.shape[0]
-#             np_mean = np_y_sum / num_examples
-
-#             np_gmrae += np.log(np.abs(np_y_i - np_y_pred_i) / np.abs(np_y_i - np_mean)).sum()
-
-#         assert pytest.approx(res, rel=1e-4) == np.exp(np_gmrae / num_examples)
-
-#     metric_devices = ["cpu"]
-#     if device.type != "xla":
-#         metric_devices.append(idist.device())
-#     for metric_device in metric_devices:
-#         for _ in range(2):
-#             _test(n_epochs=1, metric_device=metric_device)
-#             _test(n_epochs=2, metric_device=metric_device)
-
-
-# @pytest.mark.distributed
-# @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
-# @pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU")
-# def test_distrib_nccl_gpu(distributed_context_single_node_nccl):
-
-#     device = idist.device()
-#     _test_distrib_compute(device)
-#     _test_distrib_integration(device)
-
-
-# @pytest.mark.distributed
-# @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
-# def test_distrib_gloo_cpu_or_gpu(distributed_context_single_node_gloo):
-
-#     device = idist.device()
-#     _test_distrib_compute(device)
-#     _test_distrib_integration(device)
-
-
-# @pytest.mark.distributed
-# @pytest.mark.skipif(not idist.has_hvd_support, reason="Skip if no Horovod dist support")
-# @pytest.mark.skipif("WORLD_SIZE" in os.environ, reason="Skip if launched as multiproc")
-# def test_distrib_hvd(gloo_hvd_executor):
-#     device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
-#     nproc = 4 if not torch.cuda.is_available() else torch.cuda.device_count()
-
-#     gloo_hvd_executor(_test_distrib_compute, (device,), np=nproc, do_init=True)
-#     gloo_hvd_executor(_test_distrib_integration, (device,), np=nproc, do_init=True)
-
-
-# @pytest.mark.multinode_distributed
-# @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
-# @pytest.mark.skipif("MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
-# def test_multinode_distrib_gloo_cpu_or_gpu(distributed_context_multi_node_gloo):
-
-#     device = idist.device()
-#     _test_distrib_compute(device)
-#     _test_distrib_integration(device)
-
-
-# @pytest.mark.multinode_distributed
-# @pytest.mark.skipif(not idist.has_native_dist_support, reason="Skip if no native dist support")
-# @pytest.mark.skipif("GPU_MULTINODE_DISTRIB" not in os.environ, reason="Skip if not multi-node distributed")
-# def test_multinode_distrib_nccl_gpu(distributed_context_multi_node_nccl):
-
-#     device = idist.device()
-#     _test_distrib_compute(device)
-
-
-# @pytest.mark.tpu
-# @pytest.mark.skipif("NUM_TPU_WORKERS" in os.environ, reason="Skip if NUM_TPU_WORKERS is in env vars")
-# @pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
-# def test_distrib_single_device_xla():
-#     device = idist.device()
-#     _test_distrib_compute(device)
-#     _test_distrib_integration(device)
-
-
-# def _test_distrib_xla_nprocs(index):
-#     device = idist.device()
-#     _test_distrib_compute(device)
-#     _test_distrib_integration(device)
-
-
-# @pytest.mark.tpu
-# @pytest.mark.skipif("NUM_TPU_WORKERS" not in os.environ, reason="Skip if no NUM_TPU_WORKERS in env vars")
-# @pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
-# def test_distrib_xla_nprocs(xmp_executor):
-#     n = int(os.environ["NUM_TPU_WORKERS"])
-#     xmp_executor(_test_distrib_xla_nprocs, args=(), nprocs=n)
+@pytest.mark.tpu
+@pytest.mark.skipif("NUM_TPU_WORKERS" not in os.environ, reason="Skip if no NUM_TPU_WORKERS in env vars")
+@pytest.mark.skipif(not idist.has_xla_support, reason="Skip if no PyTorch XLA package")
+def test_distrib_xla_nprocs(xmp_executor):
+    n = int(os.environ["NUM_TPU_WORKERS"])
+    xmp_executor(_test_distrib_xla_nprocs, args=(), nprocs=n)

